x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/InferenceStats.java

```diff
@@ -0,0 +1,268 @@
+        NAME,
+        true,
+        a -> new InferenceStats((Long)a[0], (Long)a[1], (Long)a[2], (Long)a[3], (String)a[4], (String)a[5], (Instant) a[6])
+    );
+    static {
+        PARSER.declareLong(ConstructingObjectParser.constructorArg(), MISSING_ALL_FIELDS_COUNT);
+        PARSER.declareLong(ConstructingObjectParser.constructorArg(), INFERENCE_COUNT);
+        PARSER.declareLong(ConstructingObjectParser.constructorArg(), TOTAL_TIME_SPENT_MILLIS);
+        PARSER.declareLong(ConstructingObjectParser.constructorArg(), FAILURE_COUNT);
+        PARSER.declareString(ConstructingObjectParser.constructorArg(), MODEL_ID);
+        PARSER.declareString(ConstructingObjectParser.constructorArg(), NODE_ID);
+        PARSER.declareField(ConstructingObjectParser.constructorArg(),
+            p -> TimeUtils.parseTimeFieldToInstant(p, TIMESTAMP.getPreferredName()),
+            TIMESTAMP,
+            ObjectParser.ValueType.VALUE);
+    }
+    public static InferenceStats emptyStats(String modelId, String nodeId) {
+        return new InferenceStats(0L, 0L, 0L, 0L, modelId, nodeId, Instant.now());
+    }
+
+    public static String docId(String modelId, String nodeId) {
+        return NAME + "-" + modelId + "-" + nodeId;
+    }
+
+    private final long missingAllFieldsCount;
+    private final long inferenceCount;
+    private final long totalTimeSpent;
+    private final long failureCount;
+    private final String modelId;
+    private final String nodeId;
+    private final Instant timeStamp;
+
+    private InferenceStats(Long missingAllFieldsCount,
+                           Long inferenceCount,
+                           Long totalTimeSpent,
+                           Long failureCount,
+                           String modelId,
+                           String nodeId,
+                           Instant instant) {
+        this(unbox(missingAllFieldsCount),
+            unbox(inferenceCount),
+            unbox(totalTimeSpent),
+            unbox(failureCount),
+            modelId,
+            nodeId,
+            instant == null ? Instant.now() : Instant.ofEpochMilli(instant.toEpochMilli()));
+    }
+
+
+    public InferenceStats(long missingAllFieldsCount,
+                          long inferenceCount,
+                          long totalTimeSpent,
+                          long failureCount,
+                          String modelId,
+                          String nodeId,
+                          Instant timeStamp) {
+        this.missingAllFieldsCount = missingAllFieldsCount;
+        this.inferenceCount = inferenceCount;
+        this.totalTimeSpent = totalTimeSpent;
+        this.failureCount = failureCount;
+        this.modelId = modelId;
+        this.nodeId = nodeId;
+        this.timeStamp = timeStamp == null ?
+            Instant.ofEpochMilli(Instant.now().toEpochMilli()) :
+            Instant.ofEpochMilli(timeStamp.toEpochMilli());
```

Comment:
It's because the convention is for the XContent representation to be at millisecond accuracy and it's painful if the internal representation is more accurate.  As I said in https://github.com/elastic/elasticsearch/pull/52778#discussion_r384454026 I think it's worth adding a comment to say this.  We used to have about 3 different ways of solving this problem but now we're standardising on rounding in the constructor.

Reply:
Rounding in the ctor is the standard way. I will remove the rounding from the stream input :)